src/Assessment/Processor/ProcessDatasets.cs

```diff
@@ -1243,6 +1249,277 @@ private void UpdateIndividualWebAppPaaSCosts(List<WebApp_PaaS> WebApp_PaaS_List)
+                        EnvironmentToUniqueIdMap["Prod"].Add(sqlinstance.MachineId);
+                }
+                EnvironmentCountDictionary[sqlinstance.Environment] = +1;
+            }
+
+            SQLIaaSObj.DevMachinesCountTarget = EnvironmentCountDictionary["Dev"];
+            SQLIaaSObj.ProdMachinesCountTarget = EnvironmentCountDictionary["Prod"];
+            SQLIaaSObj.DevUniqueMachinesSource = EnvironmentToUniqueIdMap["Dev"].Count;
+            SQLIaaSObj.ProdUniqueMachinesSource = EnvironmentToUniqueIdMap["Prod"].Count;
+            SQLIaaSObj.EnvironmentToUniqueIdMap = EnvironmentToUniqueIdMap;
+            SQLIaaSObj.EnvironmentCountDictionary = EnvironmentCountDictionary;
+            return SQLIaaSObj;
+        }
+
+        private VMServer_IaaS_FS DetailsVMServer_IaaS(List<VM_IaaS_Server_Rehost_Perf> VM_IaaS_Server_Rehost_Perf_List)
+        {
+            VMServer_IaaS_FS VMServer_IaaS_FSObj = new VMServer_IaaS_FS();
+            Dictionary<string, int> EnvironmentCountDictionary = new Dictionary<string, int>();
+            Dictionary<string, HashSet<string>> EnvironmentToUniqueIdMap = new Dictionary<string, HashSet<string>>();
+            EnvironmentToUniqueIdMap["Dev"] = new HashSet<string>();
+            EnvironmentToUniqueIdMap["Prod"] = new HashSet<string>();
+
+            foreach (var machine in VM_IaaS_Server_Rehost_Perf_List)
+            {
+                if (machine.Environment.Equals("Dev"))
+                {
+                    if (!EnvironmentCountDictionary.ContainsKey("Dev"))
+                    {
+                        EnvironmentCountDictionary.Add("Dev", 0);
+                    }
+                    if (!EnvironmentToUniqueIdMap["Dev"].Contains(machine.MachineId))
+                        EnvironmentToUniqueIdMap["Dev"].Add(machine.MachineId);
+                }
+                else
+                {
+                    if (!EnvironmentCountDictionary.ContainsKey("Prod"))
+                    {
+                        EnvironmentCountDictionary.Add("Prod", 0);
+                    }
+                    if (!EnvironmentToUniqueIdMap["Prod"].Contains(machine.MachineId))
+                        EnvironmentToUniqueIdMap["Prod"].Add(machine.MachineId);
+                }
+                EnvironmentCountDictionary[machine.Environment] = +1;
+            }
+
+            VMServer_IaaS_FSObj.DevMachinesCountTarget = EnvironmentCountDictionary["Dev"];
+            VMServer_IaaS_FSObj.ProdMachinesCountTarget = EnvironmentCountDictionary["Prod"];
+            VMServer_IaaS_FSObj.DevUniqueMachinesSource = EnvironmentToUniqueIdMap["Dev"].Count;
+            VMServer_IaaS_FSObj.ProdUniqueMachinesSource = EnvironmentToUniqueIdMap["Prod"].Count;
+            VMServer_IaaS_FSObj.EnvironmentToUniqueIdMap = EnvironmentToUniqueIdMap;
+            VMServer_IaaS_FSObj.EnvironmentCountDictionary = EnvironmentCountDictionary;
+            return VMServer_IaaS_FSObj;
+        }

```

Comment:
P1: If this logic can be moved to AzureIaaSCalculator, we can save looping over the same dataset again and again.